### PR TITLE
56 update integration best practices for cloud recording, media push, media pull

### PR DESCRIPTION
**Pull request by saudsami**
Resolved the list of issues mentioned in https://github.com/AgoraIO/Doc-Source-Private/issues/56

### CODE REVIEW

File: cloud-recording/develop/integration-best-practices.md [hunk 1]
```diff
@@ -80,14 +81,48 @@ For example, if there is a fixed 5-minute break in each class, you can set `maxI
...
+
+## Integration requirements checklist
+
+To ensure reliability of the cloud recording service, refer to the following checklist to confirm that your solution meets the integration requirements:
+
+| Serial | Importance | Item | Description |
+|--|----------|----------|----------|
+| 1 | required | Subscribe to a service | Make sure you have activated the cloud recording service. | 
+| 2 | required | request method | <ul><li>To query, use the `POST` request method; to query the recording status use `GET`. </li> <li>Request URLs and request body content are case-sensitive.</li></ul> |
+| 3 | required | Get recording resources | <ul><li>The passed in `uid` cannot duplicate any UID within the current channel.</li><li>For page recording, make sure an appid + cname + uid corresponds to a resource ID.</li><li>Make sure that a resource ID is only used for one cloud recording service.</li><li>Make sure to call the `start` method.</li></ul> |
```
Comment (atovpeko) [inline cloud-recording/develop/integration-best-practices.md:98]: I'd say appid + cname + uid  should be in code font

File: cloud-recording/develop/integration-best-practices.md [hunk 2]
```diff
@@ -80,14 +81,48 @@ For example, if there is a fixed 5-minute break in each class, you can set `maxI
...
+| 4 | required | channel scene | Make sure that the channel scene (`channelType`) is consistent with the settings of the <Vg k="VSDK" />. |
+| 5 | required | recording parameters | <ul><li>Make sure that the type, case, and value range of all the parameters passed in when starting the recording are correct, and the required parameters are filled; otherwise, error code 2 is returned.</li><li>Set the layout and video bit rate of the combined recording by referring to the combined layout document and the recording bit rate comparison table.</li></ul> |
+| 6 | required | Confirm that the recording service has started successfully | <ul><li>Make sure that the start request is successful, that is, the `sid` (recording ID) is successfully obtained.</li><li>Make sure to be or within 90 seconds `sid` of getting status 4 5</li></ul> |
```
Comment (atovpeko) [inline cloud-recording/develop/integration-best-practices.md:101]: Can this be rephrased to make it clearer? 

> Make sure to be or within 90 seconds `sid` of getting status 4 5

File: cloud-recording/develop/integration-best-practices.md [hunk 3]
```diff
@@ -80,14 +81,48 @@ For example, if there is a fixed 5-minute break in each class, you can set `maxI
...
+| 7 | required | PCW & QPS limits | <ul><li>Make sure that each App ID does not exceed 10 requests per second (QPS).</li><li>Ensure that the maximum number of concurrent tasks (PCW) in China does not exceed 1,000, and that in other regions does not exceed 300. To increase the QPS and PCW limits, please contact technical support.</li></ul> |
```
Comment (atovpeko) [inline cloud-recording/develop/integration-best-practices.md:102]:
```suggestion
| 7 | required | PCW & QPS limits | <ul><li>Make sure that each App ID does not exceed 10 requests per second (QPS).</li><li>Ensure that the maximum number of concurrent tasks (PCW) in China does not exceed 1,000, and that in other regions it does not exceed 300. To increase the QPS and PCW limits, please contact technical support.</li></ul> |
```

File: cloud-recording/develop/integration-best-practices.md [hunk 4]
```diff
@@ -80,14 +81,48 @@ For example, if there is a fixed 5-minute break in each class, you can set `maxI
...
+| 8 | optional | NCS Service Activation | Activate the cloud recording callback service and subscribe to the following events as an auxiliary means of monitoring the recording service status: <ul><li>`40 recorder_started`: The recording service has started.</li><li>`11 session_exit`: The recording service ended the task and exited.</li><li>`1 cloud_recording_error`: An error occurred in the recording service.</li><li>`12 session_failover`: Enable high availability mechanism for recording.</li><li>`31 uploaded`: All recorded files have been uploaded to the specified third-party cloud storage.</li></ul> |
+| 9 | optional | Use dual domain names | If the request fails with the primary domain name `api.agora.io`, try again with the primary domain name; if it fails again, switch to the secondary domain name `api.sd-rtn.com` and send the request again. |
```
Comment (atovpeko) [inline cloud-recording/develop/integration-best-practices.md:104]: Google style guide advises to use semi-colons sparingly. I would say in this case, these can be two separate sentences.

File: media-push/develop/integration-best-practices.mdx [hunk 5]
```diff
@@ -73,105 +71,42 @@ the following two ways to implement it:
...
--   If the current primary domain name is `api.agora.io`, the alternate
-    domain name is `api.sd-rtn.com`.
-
-4. If the retry in step 3 still fails, use the domain name adjacent to
-the current region to retry.
-
-For example, suppose your business server is located in Europe. You set the primary domain name to `api.agora.io`, and
-the business server resolves the primary domain name to Germany. Germany is located in Central Europe (`api-eu-central-1.agora.io`).
-Check the domain name table below - the adjacent area is West Europe (`api-eu-west-1.agora.io` or `api-eu-west-1.sd-rtn.com`).
+For use cases with higher quality requirements, requesting multiple Media push tasks with different tagging enables isolated resources for these tasks and raises the success rate. Media push supports room-level isolation for tasks. For example, if there is a task with `"name":"my-task"` in the original request, you can add a suffix to the `name` parameter to realize isolation at the computer room level. Use `"name":"my-task-agora-isolate-1"` and `"name":"my-task-agora-isolate-2"` to start additional tasks. Note that:
```
Comment (atovpeko) [inline media-push/develop/integration-best-practices.mdx:74]: media push tasks but Media Push ( to be consistent)

File: media-push/develop/integration-best-practices.mdx [hunk 6]
```diff
@@ -73,105 +71,42 @@ the following two ways to implement it:
...
-[`api.agora.io`](`http://api.agora.io`)       |[`api-us-west-1.agora.io`](`http://api-us-west-1.agora.io`)             |Western United States |
-|[`api-us-east-1.agora.io`](`http://api-us-east-1.agora.io`)               |Eastern United States |
-|[`api-ap-southeast-1.agora.io`](`http://api-ap-southeast-1.agora.io`)     |Southeast Asia Pacific|
-|[`api-ap-northeast-1.agora.io`](`http://api-ap-northeast-1.agora.io`)     |Northeast Asia Pacific|
-|[`api-eu-west-1.agora.io`](`http://api-eu-west-1.agora.io`)               |Western Europe        |
-|[`api-eu-central-1.agora.io`](`http://api-eu-central-1.agora.io`)         |Central Europe        |
-|[`api-cn-east-1.agora.io`](`http://api-cn-east-1.agora.io`)               |East China            |
-|[`api-cn-north-1.agora.io`](`http://api-cn-north-1.agora.io`)             |North China           |
+* If resources are insufficient, then tasks started later will fail.
+* Media push supports up to three-way isolation, That is, `my-taskagora-isolate1`,`my-task-agoraisolate-2`, and `my-taskagora-isolate-3` can be separated.
```
Comment (atovpeko) [inline media-push/develop/integration-best-practices.mdx:77]:
```suggestion
* Media Push supports up to three-way isolation, That is, `my-taskagora-isolate1`, `my-task-agoraisolate-2`, and `my-taskagora-isolate-3` can be separated.
```

File: media-push/develop/integration-best-practices.mdx [hunk 7]
```diff
@@ -73,105 +71,42 @@ the following two ways to implement it:
...
+<SwitchDomainName/>
 
 ## Create RTMP Converter
 
 When creating a converter with a `create` call, pay attention to the following:
 
 - Create the converter by region parameter:
-    - The region you set must be in the same area as your CDN origin station. For example, if your CDN origin is located in US,
-    set region to `na`.
+    - The region you set must be in the same area as your CDN origin station. For example, if your CDN origin is located in US, set region to `na`.
```
Comment (atovpeko) [inline media-push/develop/integration-best-practices.mdx:86]:
```suggestion
    - The region you set must be in the same area as your CDN origin station. For example, if your CDN origin is located in the US, set region to `na`.
```

File: media-push/develop/integration-best-practices.mdx [hunk 8]
```diff
@@ -73,105 +71,42 @@ the following two ways to implement it:
...
-`show68_horizontal` and `show68_vertical` would represent a horizontal and a vertical layout respectively.
+- Agora recommends that you assign a string value to the `X-Request-ID` field in the request header. The Agora server returns an `X-Custom-request-ID` field in the response header for troubleshooting purposes.
+- Set UID or account as the user name for the converter. Do not set these two fields at the same time. Ensure that each converter has a unique user name within the channel.
+- To avoid repeated media streaming due to repeated creation of multiple converters, it is recommended to use the name field to manage converters under a specific project.    At the same time, converters with the same names cannot exist in a project. If you attempt to create a converter with the same name, you will receive the `409` (Conflict) status code.
+-   Agora recommends that you assign the name using a combination of the channel name and a converter property. For example, `show68_horizontal` and `show68_vertical` would represent a horizontal and a vertical layout respectively.
```
Comment (atovpeko) [inline media-push/develop/integration-best-practices.mdx:91]: These documents use "will" multiple times. However, future tense in general should be avoided: https://developers.google.com/style/tense

File: media-push/develop/integration-best-practices.mdx [hunk 9]
```diff
@@ -73,105 +71,42 @@ the following two ways to implement it:
...
 -   `audioOption` and `videoOptions` are set as follows for audio only or video only scenarios:
     - In a video only scenario, you do not need to set `audioOptions` and its related fields.
     - In a audio only scenario, you do not need to set `videoOptions` and their related fields.
-    - In audio plus video scenarios, `videoOptions` and `audioOptions` are required and cannot be left blank.
-    > blank.
--   Set an appropriate value for `idleTimeout`. The default value of 300 seconds is recommended. It means that the converter
-will automatically destroy 300 seconds after all the subscribed leave the channel.
+    - In audio plus video scenarios, `videoOptions` and `audioOptions` are required and cannot be left blank. If you do not need to configure `audioOptions` set it to null.
```
Comment (atovpeko) [inline media-push/develop/integration-best-practices.mdx:95]:
```suggestion
    - In audio plus video scenarios, `videoOptions` and `audioOptions` are required and cannot be left blank. If you do not need to configure `audioOptions` set it to `null`.
```

File: media-push/develop/integration-best-practices.mdx [hunk 10]
```diff
@@ -185,12 +120,9 @@ Update API does not support the following configuration of the update `Converter
 
 ### Retry with retreat strategy
 
-If you receive `404`, `429`, or `5xx` error codes, take an escape strategy. For example, wait 5-6 seconds, 10-11 seconds,
-or 15-16 seconds and try again. If you receive `404` and `5xx` three times in a row or if the `state` field in the body of the
-response to querying the push status is `failed`, then the media push task has failed. Perform the following steps to rectify the fault:
+If you receive `404`, `429`, or `5xx` error codes, take an escape strategy. For example, wait 5-6 seconds, 10-11 seconds, or 15-16 seconds and try again. If you get three consecutive `404` or `5xx` error codes, or if the `state` field in the response to checking the status of a media push task says `failed`, it means that the media push task has failed. To fix this problem, follow these steps:
```
Comment (atovpeko) [inline media-push/develop/integration-best-practices.mdx:123]: I think backoff/escape/retreat strategy is the same thing - could we make sure to be consistent?

File: shared/common/_switch-domain-name.mdx [hunk 11]
```diff
@@ -0,0 +1,43 @@
+
+### Switch domain name
+
+Take the following steps to switch a domain name:
+
+1. Set the primary domain name based on the location of your service server:
+    - If the DNS address of the service server is located in a country or region other than mainland China, set the primary domain name to `api.agora.io`.
+    - If the DNS address of the service server is in mainland China, set the primary domain name to `api.sd-rtn.com`.
+2. Setup your retry strategy in the following order:
```
Comment (atovpeko) [inline shared/common/_switch-domain-name.mdx:9]:
```suggestion
2. Set up your retry strategy in the following order:
```

File: shared/common/_switch-domain-name.mdx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        - If the current primary domain name is `api.agora.io`, the alternate domain name is `api.sd-rtn.com`.
+    
+    1. **Adjacent domain retry**: If alternate domain retry fails, retry using the domain name adjacent to the current region.
+
+        For example, suppose your business server is located in Europe. You set the primary domain name to `api.agora.io`, and the business server resolves the primary domain name to Germany. Germany is located in central Europe (`api-eu-central-1.agora.io`). The [domain name table](#domain-name-table) shows that the adjacent area is West Europe. Use the `api-eu-west-1.agora.io` or `api-eu-west-1.sd-rtn.com` domain name to retry.
+
+
+To avoid exceeding the QPS limit with retry requests, best practice is to do your first retry after 1 second, then 3 seconds, then 6 seconds. If the request fails because of a network problem rather than a DNS domain name resolution problem, skip Alternate domain retry and proceed to Adjacent domain retry. Before switching to the region domain name, ensure that the REST services you use, for example, cloud recording or channel management, are deployed in that region.
```
Comment (atovpeko) [inline shared/common/_switch-domain-name.mdx:22]: why capitalize Alternate and Adjacent?

**Review by atovpeko — CHANGES_REQUESTED** [review]

File: cloud-recording/develop/integration-best-practices.md (hunk 1 above)
Comment (saudsami) [reply cloud-recording/develop/integration-best-practices.md:98]: fixed

File: cloud-recording/develop/integration-best-practices.md (hunk 2 above)
Comment (saudsami) [reply cloud-recording/develop/integration-best-practices.md:101]: Fixed

File: cloud-recording/develop/integration-best-practices.md (hunk 3 above)
Comment (saudsami) [reply cloud-recording/develop/integration-best-practices.md:102]: fixed

File: cloud-recording/develop/integration-best-practices.md (hunk 4 above)
Comment (saudsami) [reply cloud-recording/develop/integration-best-practices.md:104]: Done.

File: media-push/develop/integration-best-practices.mdx (hunk 5 above)
Comment (saudsami) [reply media-push/develop/integration-best-practices.mdx:74]: Changed all occurrences to Media Push

File: media-push/develop/integration-best-practices.mdx (hunk 6 above)
Comment (saudsami) [reply media-push/develop/integration-best-practices.mdx:77]: Fixed

File: media-push/develop/integration-best-practices.mdx (hunk 7 above)
Comment (saudsami) [reply media-push/develop/integration-best-practices.mdx:86]: fixed

File: media-push/develop/integration-best-practices.mdx (hunk 8 above)
Comment (saudsami) [reply media-push/develop/integration-best-practices.mdx:91]: Removed all occurrences of "will"

File: media-push/develop/integration-best-practices.mdx (hunk 10 above)
Comment (saudsami) [reply media-push/develop/integration-best-practices.mdx:123]: Changed escape to backoff

File: shared/common/_switch-domain-name.mdx (hunk 12 above)
Comment (saudsami) [reply shared/common/_switch-domain-name.mdx:22]: fixed

**Review by billy-the-fish — APPROVED** [review]
LGTM.

**Pull request merged**